### PR TITLE
HPCC-15439 Implement a single permissions cache shared by all sec mgrs

### PR DESCRIPTION
**Pull request by RussWhitehead**
Implement a new "sharedCache" config option that controls if the security
manager cache is share between all instances, or is private to a each
instance.  Also create a new singleton cache factory that returns a single
shared instance of the cache if the sharedCache option is chosen

Signed-off-by: Russ Whitehead william.whitehead@lexisnexis.com

### CODE REVIEW

**Comment by RussWhitehead** [comment]
@richardkchapman Please consider merging this into 6.0.0

File: system/security/shared/caching.hpp [hunk 1]
```diff
@@ -134,19 +134,36 @@ class CachedUser
...
         m_secMgr = NULL;
         m_lastManagedFileScopesRefresh = 0;
         m_defaultPermission = SecAccess_Unknown;
     }
+
     virtual ~CPermissionsCache();
 
+    static CPermissionsCache* queryInstance()
+    {
+        static CPermissionsCache* instance = nullptr;
```
Comment (richardkchapman) [inline system/security/shared/caching.hpp]: More efficient to move this out of the method

**Comment by richardkchapman** [comment]
@RussWhitehead One minor quibble, otherwise looks clean enough

**Comment by RussWhitehead** [comment]
Moved the static instance member to global file scope. @richardkchapman Please rereview

**Comment by richardkchapman** [comment]
Your second commit seems to just have the signature with no title. If you had the git hooks that check commit validity in place it should prohibit this.

I will squash and merge manually.

**Comment by richardkchapman** [comment]
Merged manually into 6.0.0

**Pull request closed without merging**